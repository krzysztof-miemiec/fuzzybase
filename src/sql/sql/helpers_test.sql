CREATE EXTENSION fuzzy;

-- Fuzzy helpers
SELECT add_fuzzy_function('heart_rate', '0', 'none')
UNION SELECT add_fuzzy_function('heart_rate', '1~35\50', 'very low')
UNION SELECT add_fuzzy_function('heart_rate', '35/40~60\70', 'low')
UNION SELECT add_fuzzy_function('heart_rate', '50/60~100\130', 'normal')
UNION SELECT add_fuzzy_function('heart_rate', '120/130~160\190', 'high')
UNION SELECT add_fuzzy_function('heart_rate', '160/180~300', 'very high')
UNION SELECT add_fuzzy_function('heart_rate', '250/300~600\900', 'extremely high')
UNION SELECT add_fuzzy_function('heart_rate', '300/400~1200', 'impossible');

-- Creation and selection
SELECT x, get_fuzzy_name('heart_rate', x) FROM generate_series(10, 220, 50) AS x;

-- Deletion
SELECT delete_fuzzy_function('heart_rate', 'normal');
SELECT delete_fuzzy_function('heart_rate', trapezoidal_function '0');
SELECT * FROM get_fuzzy_type_functions('heart_rate');

-- Deletion of the whole type
SELECT delete_fuzzy_type('heart_rate');
SELECT * FROM get_fuzzy_type_functions('heart_rate');

-- Creation of a single type
SELECT add_fuzzy_type('cholesterol');
SELECT add_fuzzy_function('heart_rate', '35/40~60\70', 'low')
UNION SELECT add_fuzzy_function('heart_rate', '50/60~100\130', 'normal')
UNION SELECT add_fuzzy_function('heart_rate', '120/130~160\190', 'high');

-- Selection of all types
SELECT * FROM get_fuzzy_types();

-- Getter for a specific function
SELECT * FROM get_fuzzy_function('heart_rate', 'high');
SELECT 'heart_rate'|>'high';
SELECT  130~>'heart_rate';

DROP EXTENSION fuzzy CASCADE;
