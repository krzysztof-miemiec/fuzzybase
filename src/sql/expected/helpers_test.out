CREATE EXTENSION fuzzy;
-- Fuzzy helpers
SELECT add_fuzzy_function('heart_rate', '0', 'none')
UNION SELECT add_fuzzy_function('heart_rate', '1~35\50', 'very low')
UNION SELECT add_fuzzy_function('heart_rate', '35/40~60\70', 'low')
UNION SELECT add_fuzzy_function('heart_rate', '50/60~100\130', 'normal')
UNION SELECT add_fuzzy_function('heart_rate', '120/130~160\190', 'high')
UNION SELECT add_fuzzy_function('heart_rate', '160/180~300', 'very high')
UNION SELECT add_fuzzy_function('heart_rate', '250/300~600\900', 'extremely high')
UNION SELECT add_fuzzy_function('heart_rate', '300/400~1200', 'impossible');
                 add_fuzzy_function                  
-----------------------------------------------------
 (1,"120.00/130.00~160.00\\190.00",high)
 (1,"0.00/0.00~0.00\\0.00",none)
 (1,"300.00/400.00~1200.00\\1200.00",impossible)
 (1,"250.00/300.00~600.00\\900.00","extremely high")
 (1,"160.00/180.00~300.00\\300.00","very high")
 (1,"1.00/1.00~35.00\\50.00","very low")
 (1,"50.00/60.00~100.00\\130.00",normal)
 (1,"35.00/40.00~60.00\\70.00",low)
(8 rows)

-- Creation and selection
SELECT x, get_fuzzy_name('heart_rate', x) FROM generate_series(10, 220, 50) AS x;
  x  | get_fuzzy_name 
-----+----------------
  10 | very low
  60 | low
 110 | normal
 160 | high
 210 | very high
(5 rows)

-- Deletion
SELECT delete_fuzzy_function('heart_rate', 'normal');
 delete_fuzzy_function 
-----------------------
 (1,normal)
(1 row)

SELECT delete_fuzzy_function('heart_rate', trapezoidal_function '0');
   delete_fuzzy_function    
----------------------------
 (1,"0.00/0.00~0.00\\0.00")
(1 row)

SELECT * FROM get_fuzzy_type_functions('heart_rate');
             func              |   range_name   
-------------------------------+----------------
 1.00/1.00~35.00\50.00         | very low
 35.00/40.00~60.00\70.00       | low
 120.00/130.00~160.00\190.00   | high
 160.00/180.00~300.00\300.00   | very high
 250.00/300.00~600.00\900.00   | extremely high
 300.00/400.00~1200.00\1200.00 | impossible
(6 rows)

-- Deletion of the whole type
SELECT delete_fuzzy_type('heart_rate');
 delete_fuzzy_type 
-------------------
 heart_rate
(1 row)

SELECT * FROM get_fuzzy_type_functions('heart_rate');
ERROR:  The type does not exist.
CONTEXT:  PL/pgSQL function get_fuzzy_type_functions(character varying) line 7 at RAISE
-- Creation of a single type
SELECT add_fuzzy_type('cholesterol');
 add_fuzzy_type 
----------------
              9
(1 row)

SELECT add_fuzzy_function('heart_rate', '35/40~60\70', 'low')
UNION SELECT add_fuzzy_function('heart_rate', '50/60~100\130', 'normal')
UNION SELECT add_fuzzy_function('heart_rate', '120/130~160\190', 'high');
            add_fuzzy_function            
------------------------------------------
 (10,"120.00/130.00~160.00\\190.00",high)
 (10,"50.00/60.00~100.00\\130.00",normal)
 (10,"35.00/40.00~60.00\\70.00",low)
(3 rows)

-- Selection of all types
SELECT * FROM get_fuzzy_types();
  type_name  |    range_names    
-------------+-------------------
 cholesterol | {NULL}
 heart_rate  | {low,normal,high}
(2 rows)

DROP EXTENSION fuzzy CASCADE;
